app/Models/Album.php

```diff
@@ -259,4 +271,58 @@ public function newEloquentBuilder($query): AlbumBuilder
 	{
 		return new AlbumBuilder($query);
 	}
+
+	/**
+	 * Accessor for the "virtual" attribute {@link Album::$track_url}.
+	 *
+	 * This is a convenient method which wraps
+	 * {@link Album::$track_short_path} into
+	 * {@link \Illuminate\Support\Facades\Storage::url()}.
+	 *
+	 * @return string|null the url of the track
+	 */
+	public function getTrackUrlAttribute(): ?string
+	{
+		return $this->track_short_path !== null && $this->track_short_path !== '' ?
+			Storage::url($this->track_short_path) : null;
+	}
+
+	/**
+	 * Set the GPX track for the album.
+	 *
+	 * @param UploadedFile $file the GPX track file to be set
+	 *
+	 * @return void
+	 *
+	 * @throws \Illuminate\Contracts\Filesystem\FileNotFoundException
+	 */
+	public function setTrack(UploadedFile $file): void
+	{
+		if ($this->track_short_path !== null) {
+			Storage::delete($this->track_short_path);
+		}
+
+		$new_track_id = strtr(base64_encode(random_bytes(18)), '+/', '-_');
+		Storage::putFileAs('tracks/', $file, "$new_track_id.xml");
+		$this->track_short_path = "tracks/$new_track_id.xml";
+		$this->save();
+	}
+
+	/**
+	 * Delete the track of the album.
+	 *
+	 * @return void
+	 */
+	public function deleteTrack(): void
+	{
+		if ($this->track_short_path === null) {
+			return;
+		}
+		// make sure track is not used for other albums
+		if (self::query()->where('track_short_path', '=', $this->track_short_path)->count() <= 1) {
+			Storage::delete($this->track_short_path);
+		}
+		$this->track_short_path = null;
+		$this->save();
+	}
```

Comment:
Same thing here. BTW. The query is not necessary anymore as the file names are truly random.

```suggestion
	/**
	 * Delete the track of the album.
	 *
	 * @return void
	 *
	 * @throws ModelDBException
	 */
	public function deleteTrack(): void
	{
		if ($this->track_short_path === null) {
			return;
		}
		Storage::delete($this->track_short_path);
		$this->track_short_path = null;
		$this->save();
	}
```